src/heliaServer.ts

```diff
@@ -163,58 +163,59 @@ export class HeliaServer {
-    this.log('Fetching without subdomain')
-    const opController = new AbortController()
-    setMaxListeners(Infinity, opController.signal)
-    request.raw.on('close', () => {
-      if (request.raw.aborted) {
-        this.log('Request aborted by client')
-        opController.abort()
-      }
-    })
-    // const { ns: namespace, address, '*': relativePath } = request.params as EntryParams
-    try {
-      await this.isReady
-      await this._fetch({ request, reply, address, namespace, relativePath, signal: opController.signal, peerId, cid })
-    } catch (error) {
-      this.log('Error requesting content from helia:', error)
-      await reply.code(500).send(error)
+  #convertVerifiedFetchResponseToFastifyReply = async (verifiedFetchResponse: Response, reply: FastifyReply): Promise<void> => {
+    if (!verifiedFetchResponse.ok) {
+      this.log('verified-fetch response not ok: ', verifiedFetchResponse.status)
+      await reply.code(verifiedFetchResponse.status).send(verifiedFetchResponse.statusText)
+      return
+    }
+    const contentType = verifiedFetchResponse.headers.get('content-type')
+    if (contentType == null) {
+      this.log('verified-fetch response has no content-type')
+      await reply.code(200).send(verifiedFetchResponse.body)
+      return
+    }
+    if (verifiedFetchResponse.body == null) {
+      this.log('verified-fetch response has no body')
+      await reply.code(501).send('empty')
+      return
     }
-  }
 
-  async _fetch ({ reply, address, namespace, relativePath, signal, cid, peerId }: RouteHandler & HeliaFetchOptions & { signal: AbortSignal }): Promise<void> {
-    this.log('Fetching from Helia:', { address, namespace, relativePath })
-    let type: string | undefined
-    // raw response is needed to respond with the correct content type.
+    const headers: Record<string, string> = {}
+    for (const [headerName, headerValue] of verifiedFetchResponse.headers.entries()) {
+      headers[headerName] = headerValue
+    }
+    // Fastify really does not like streams despite what the documentation and github issues say.
+    const reader = verifiedFetchResponse.body.getReader()
+    reply.raw.writeHead(verifiedFetchResponse.status, headers)
     try {
-      for await (const chunk of await this.heliaFetch.fetch({ address, namespace, relativePath, signal, cid, peerId })) {
-        if (type === undefined) {
-          type = await parseContentType({ bytes: chunk, path: relativePath })
-          // this needs to happen first.
-          reply.raw.writeHead(200, {
-            'Content-Type': type ?? DEFAULT_MIME_TYPE,
-            'Cache-Control': 'public, max-age=31536000, immutable'
-          })
+      let done = false
+      while (!done) {
+        const { done: _done, value } = await reader.read()
+        if (value != null) {
+          reply.raw.write(Buffer.from(value))
```

Comment:
This ignores backpressure. `reply.raw.write` returns a `boolean` value - if it's false we should wait for the stream to emit the `'drain'` event before continuing to write to it.